File: 02/java/MaxHeap.java

public class MaxHeap<E extends Comparable<E>> {
    private Array<E> data;

    public MaxHeap(int capacity) {
        data = new Array<>(capacity);
    }

    public MaxHeap() {
        data = new Array<>();
    }

    public MaxHeap(E[] arr) {
        data = new Array<>(arr);
        int index = parent(size() - 1);
        while( index >= 0 ) {
            siftDown(index);
            index--;
        }
    }

    public int size() {
        return data.getSize();
    }

    public boolean isEmpty() {
        return data.isEmpty();
    }

    private int parent(int index) {
        if( index == 0 ) {
            throw new IllegalArgumentException("index-0 doesn't have parent.");
        }
        return (index - 1) / 2;
    }

    private int lchild(int index) {
        return 2 * index + 1;
    }

    private int rchild(int index) {
        return 2 * index + 2;
    }

    public void add(E e) {
        data.addLast(e);
        siftUp(data.getSize() - 1);
    }

    private void siftUp(int index) {
        while(index > 0) {
            int pindex = parent(index);

            E pe = data.get(pindex);
            E se = data.get(index);

            if( se.compareTo(pe) > 0 ) {
                data.swap(index, pindex);
                index = pindex;
            } else {
                index = 0;
            }
        }
    }

    public E findMax() {
        if( data.getSize() == 0 ) {
            throw new IllegalArgumentException("Can not findMax when heap is empty.");
        }
        return data.get(0);
    }

    public E extractMax() {
        E max = findMax();
        data.swap(0, size() - 1);
        data.removeLast();

        siftDown(0);
        return max;
    }

    private void siftDown(int index) {
        int v = lchild(index);
        if( v < data.getSize() ) {
            int r = rchild(index);
            if( r < data.getSize() && data.get(v).compareTo(data.get(r)) < 0 ) {
                v = r;
            }

            if( data.get(v).compareTo(data.get(index)) > 0 ) {
                data.swap(v, index);
                siftDown(v);
            }
        }
        
        /* while( lchild(index) < data.getSize() ) {
            int v = lchild(index);
            int r = rchild(index);

            if( r < data.getSize() && data.get(v).compareTo(data.get(r)) < 0 ) {
                v = r;
            }

            if( data.get(v).compareTo(data.get(index)) > 0 ) {
                data.swap(v, index);
                index = v;
            }
        } */
    }

    public E replace(E e) {
        E max = findMax();
        data.set(0, e);

        siftDown(0);
        return max;
    }
}

File: 02/js/MaxHeap.js
const SArray = global.Array;
const Array = require('./Array.js');

class MaxHeap {
    constructor(arr) {
        if( arr && SArray.isArray(arr) ) {
            this.data = new Array(arr);
            this.heapify();
        } else {
            this.data = new Array();
        }
    }

    heapify() {
        let index = this._parent(this.size()-1);
        while( index >= 0 ) {
            this._siftDown(index);
            index--;
        }
    }

    size() { return this.data.getSize(); }
    isEmpty() { return this.data.isEmpty(); }

    _parent(index) {
        if( index === 0 ) throw new Error("index-0 doesn't have parent.");
        return Math.floor(( index - 1 ) / 2);
    }
    _lchild(index) { return index * 2 + 1; }
    _rchild(index) { return index * 2 + 2; }

    findMax() {
        if( this.size() === 0 ) throw new Error("Can not findMax when heap is empty.");
        return this.data.get(0);
    }

    add(elm) {
        this.data.addLast(elm);
        this._siftUp(this.size() - 1);
    }
    _siftUp(index) {
        while( index > 0 ) {
            const pindex = this._parent(index);
            if( this.data.get(index) > this.data.get(pindex) ) {
                this.data.swap(index, pindex);
                index = pindex;
            } else {
                index = 0;
            }
        }
    }

    extractMax() {
        const max = this.findMax();
        this.data.swap(0, this.size()-1);
        this.data.removeLast();

        this._siftDown(0);
        return max;
    }
    _siftDown(index) {
        let v = this._lchild(index);
        if( v < this.size() ) {
            const r = this._rchild(index);
            if( r < this.size() && this.data.get(r) > this.data.get(v) ) {
                v = r;
            }

            if( this.data.get(index) < this.data.get(v) ) {
                this.data.swap(index, v);
                this._siftDown(v);
            }
        }
    }

    replace(elm) {
        const max = this.findMax();
        this.data.set(0, elm);

        this._siftDown(0);
        return max;
    }
};

module.exports = MaxHeap;